test/test_mms.py

```diff
@@ -0,0 +1,198 @@
+import festim as F
+import numpy as np
+from dolfinx import fem
+import ufl
+from dolfinx.io import XDMFFile
+import mpi4py.MPI as MPI
+
+
+test_mesh_1d = F.Mesh1D(np.linspace(0, 1, 10000))
+
+
+def error_L2(u_computed, u_exact, degree_raise=3):
+    # Create higher order function space
+    degree = u_computed.function_space.ufl_element().degree()
+    family = u_computed.function_space.ufl_element().family()
+    mesh = u_computed.function_space.mesh
+    W = fem.FunctionSpace(mesh, (family, degree + degree_raise))
+    # Interpolate approximate solution
+    u_W = fem.Function(W)
+    u_W.interpolate(u_computed)
+
+    # Interpolate exact solution, special handling if exact solution
+    # is a ufl expression or a python lambda function
+    u_ex_W = fem.Function(W)
+    if isinstance(u_exact, ufl.core.expr.Expr):
+        u_expr = fem.Expression(u_exact, W.element.interpolation_points)
+        u_ex_W.interpolate(u_expr)
+    else:
+        u_ex_W.interpolate(u_exact)
+
+    # Compute the error in the higher order function space
+    e_W = fem.Function(W)
+    e_W.x.array[:] = u_W.x.array - u_ex_W.x.array
+
+    # Integrate the error
+    error = fem.form(ufl.inner(e_W, e_W) * ufl.dx)
+    error_local = fem.assemble_scalar(error)
+    error_global = mesh.comm.allreduce(error_local, op=MPI.SUM)
+    return np.sqrt(error_global)
+
+
+def test_MMS_steady_state():
+    """
+    MMS test with one mobile species at steady state
+    """
+
+    def u_exact(mod):
+        return lambda x: 1 + mod.sin(2 * mod.pi * x[0])
+
+    u_ufl = u_exact(ufl)
+    u_numpy = u_exact(np)
+
+    elements = ufl.FiniteElement("CG", test_mesh_1d.mesh.ufl_cell(), 1)
+    V = fem.FunctionSpace(test_mesh_1d.mesh, elements)
+    exact = fem.Function(V)
+    exact.interpolate(u_numpy)
+    exact_xdmf = XDMFFile(MPI.COMM_WORLD, "results/mms/exact_solution.xdmf", "w")
+    exact_xdmf.write_mesh(test_mesh_1d.mesh)
+    exact_xdmf.write_function(exact)
+
+    D_0 = 1
+    E_D = 0
+    D = D_0 * ufl.exp(-E_D / (F.k_B * 1))
+
+    my_model = F.HydrogenTransportProblem()
+    my_model.mesh = test_mesh_1d
+    my_mat = F.Material(name="mat", D_0=D_0, E_D=E_D)
+    vol = F.VolumeSubdomain1D(id=1, borders=[0, 1], material=my_mat)
+    left = F.SurfaceSubdomain1D(id=1, x=0)
+    right = F.SurfaceSubdomain1D(id=2, x=1)
+
+    my_model.subdomains = [vol, left, right]
+
+    A = F.Species("A")
+    my_model.species = [A]
+
+    my_model.temperature = 1
+
+    my_model.boundary_conditions = [
+        F.DirichletBC(subdomain=left, value=u_ufl, species=A),
+        F.DirichletBC(subdomain=right, value=u_ufl, species=A),
+    ]
+
+    f_expr = lambda x: D * 4 * np.pi**2 * np.sin(2 * np.pi * x[0])
```

Comment:
So you derived this source term by hand, correct?

This is fine for simple cases, but ultimately we want to test complex cases like space varying temperature, etc.

Therefore, we need to look at symbolic differenciation with sympy.

You can "lambdify" sympy expression very easily.